### PR TITLE
[AnalyticsVisualizer] Show Analytics as popovers

### PR DESCRIPTION
**Pull request by anipetrov**
Addresses https://github.com/artsy/eigen/issues/1127

### TODO list:

- [x] Create an ARAnalyticsProvider subclass that uses ARNotificationView to present the analytics event
- [x] Tapping the notification shows all the analytics properties
    - [x] add this functionality for the sign up splash screen - [currently, tapping on a notification while in sign up/log in mode does not display the properties] (http://cl.ly/0q2w2b312W25)
    - [x] show analytics properties modal for a tapped notification even if the previous properties modal was not dismissed - [currently, tapping notifications while another event's properties are displayed does not open new properties modals](http://cl.ly/0E0M0C2j0Q3v)
- [x] *Copy Description to Pasteboard* functionality
- [x] Create an Admin toggle for turning off/on the analytics mode, reset when it's changed
- [x] On app launch check if we are in beta && analytics mode is on, if so, add the provider
- [x] Optimize for notification clinging time (to alleviate lag of notifications when firing multiple events in a short period of time)
- [x] Add call-stack for where an event was triggered from 

ALLOY IS BOSS

### CODE REVIEW

**Comment by orta** [comment]
Programming is _/sooooo/_ interesting

b795a85df06


a468e4d7784

**Comment by orta** [comment]
e2a1c066d9b

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Warning</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td><h4>Found unstubbbed networking requests</h4>

<ul>
<li><a href="https://github.com/search?q=ARArtworkFavoritesNetworkModel+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARArtworkFavoritesNetworkModelSpec</a> in <code>test_getFavorites__makes_request_if_no_request_is_in_progress</code></li>
<li><a href="https://github.com/search?q=ARArtworkViewControllerBuyButton+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARArtworkViewControllerBuyButtonSpec</a> in <code>test_buy_button__posts_order_if_artwork_has_no_edition_sets</code></li>
<li><a href="https://github.com/search?q=ARCreateAccountViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARCreateAccountViewControllerSpec</a> in <code>test_warning_view__displays_a_message</code></li>
<li><a href="https://github.com/search?q=ARCreateAccountViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARCreateAccountViewControllerSpec</a> in <code>test_warning_view__displays_a_message</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__shows_favorite_and_follow_buttons_for_logged_in_user</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__shows_favorite_and_follow_buttons_for_logged_in_user</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__passes_artwork_to_share_controller</code></li>
<li><a href="https://github.com/search?q=ARFairViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARFairViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARGeneViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARGeneViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARGeneViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARGeneViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARTopMenuViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARTopMenuViewControllerSpec</a> in <code>test_navigation__favorites__logged_in__tab_behavior__already_selected__pops_to_root</code></li>
</ul>
</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by orta** [comment]
23ebc9a9edf

Check out the Empire State building!

**Comment by orta** [comment]
the Danger fail is fine, it's this https://github.com/danger/danger/issues/88

File: Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m [hunk 1]
```diff
@@ -161,6 +162,26 @@ - (ARCellData *)generateQuicksilver
...
+- (ARCellData *)generateOnScreenAnalytics
+{
+    ARCellData *crashCellData = [[ARCellData alloc] initWithIdentifier:AROptionCell];
+    [crashCellData setCellConfigurationBlock:^(UITableViewCell *cell) {
+        if ([AROptions boolForOption:AROptionsShowAnalyticsOnScreen]) {
+            cell.textLabel.text = @"Stop Analytics as Popovers";
+        } else {
+            cell.textLabel.text = @"Start Analytics as Popovers";
+        }
+    }];
```
Comment (alloy) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: Is this not using a checkmark to display the enabled status?

File: Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m [hunk 2]
```diff
@@ -161,6 +162,26 @@ - (ARCellData *)generateQuicksilver
...
+
+    [crashCellData setCellSelectionBlock:^(UITableView *tableView, NSIndexPath *indexPath) {
+        BOOL current = [AROptions boolForOption:AROptionsShowAnalyticsOnScreen];
+        [AROptions setBool:!current forOption:AROptionsShowAnalyticsOnScreen];
+        exit(YES);
```
Comment (alloy) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: I know this is already used in other places, but what is with the `YES` argument to `exit()` ?

File: Artsy/App/ARAnalyticsVisualizer.m [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- (void)event:(NSString *)event withProperties:(NSDictionary *)properties
+{
+    NSString *title = [self alertTitleForEvent:event withProperties:properties];
+
+    [ARNotificationView showNoticeInView:[self findVisibleWindow] title:title time:1.5 response:^{
+        UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:[properties description] preferredStyle:UIAlertControllerStyleActionSheet];
+        
+        [alert addAction:[UIAlertAction actionWithTitle:@"Copy Description" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
+
+
```
Comment (alloy) [inline Artsy/App/ARAnalyticsVisualizer.m]: Extraneous whitespace.

File: Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m (hunk 2 above)
Comment (alloy) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: Oh and why does the app need to be quit when changing this option?

**Comment by alloy** [comment]
Just some minor comments, otherwise looks great.

Great work, @anipetrov :clap:

File: Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m (hunk 2 above)
Comment (orta) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: the visualize is added at app launch, don't see a need to add extra code to add/remove the provider at runtime

Comment (orta) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: you have to have something in the `exit`, may as well be `YES` - think it's the exit code, so it just returns the int 1 - not really too bothered, as I think iOS just ignores it

Comment (alloy) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: > the visualize is added at app launch, don't see a need to add extra code to add/remove the provider at runtime

Ah yes, good point.

**Comment by orta** [comment]
Cool, shipped fixes

File: Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m (hunk 2 above)
Comment (alloy) [inline Artsy/View_Controllers/Admin/ARAdminSettingsViewController.m]: > you have to have something in the `exit`, may as well be `YES` - think it's the exit code, so it just returns the int 1

Hmm, it’s just a bit confusing to me, as using `YES` seems so specific and in this case case sound like “yes, no problem”, where exiting with `!0` means it was not successful, etc etc :)

> not really too bothered, as I think iOS just ignores it

Sure, it was more me wondering if there was in fact a reason for this specific exit status that I did not consider.

**Comment by orta** [comment]
Fail is danger, and known.

**Comment by anipetrov** [comment]
@alloy @orta you guys are rockstars!!

**Comment by neilkimmett** [comment]
lolcommits 🙌

**Pull request merged**